{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GitOps Flow with MLRun"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup Project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> 2023-09-18 23:36:57,974 [info] Project loaded successfully: {'project_name': 'cicd-flow'}\n",
      "Exporting project as zip archive to v3io:///bigdata/cicd-flow.zip...\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import yaml\n",
    "from src import create_and_set_project, load_config, TrainConfig, EvaluateConfig, DeployConfig\n",
    "\n",
    "\n",
    "project = create_and_set_project(\n",
    "    name=\"cicd-flow\",\n",
    "    git_source=\"v3io:///bigdata/cicd-flow.zip\",\n",
    ")\n",
    "\n",
    "config = load_config()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Training Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_config = TrainConfig(**config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>Pipeline running (id=14715a49-ad72-40c5-a584-140fc265e747), <a href=\"https://dashboard.default-tenant.app.us-tmp-353.iguazio-cd2.com/mlprojects/cicd-flow/jobs/monitor-workflows/workflow/14715a49-ad72-40c5-a584-140fc265e747\" target=\"_blank\"><b>click here</b></a> to view the details in MLRun UI</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.43.0 (0)\n",
       " -->\n",
       "<!-- Title: kfp Pages: 1 -->\n",
       "<svg width=\"622pt\" height=\"332pt\"\n",
       " viewBox=\"0.00 0.00 622.33 332.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 328)\">\n",
       "<title>kfp</title>\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-328 618.33,-328 618.33,4 -4,4\"/>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;14175956 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;14175956</title>\n",
       "<ellipse fill=\"lightgrey\" stroke=\"black\" cx=\"364.05\" cy=\"-18\" rx=\"33.29\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"364.05\" y=\"-14.3\" font-family=\"Times,serif\" font-size=\"14.00\">train</text>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;1947655188 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;1947655188</title>\n",
       "<ellipse fill=\"lightgrey\" stroke=\"black\" cx=\"457.05\" cy=\"-90\" rx=\"157.07\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"457.05\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">validate&#45;validate&#45;train&#45;test&#45;split</text>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;1947655188&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;14175956 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;1947655188&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;14175956</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M434.53,-72.05C421.36,-62.14 404.64,-49.55 390.79,-39.13\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"392.69,-36.18 382.6,-32.97 388.49,-41.78 392.69,-36.18\"/>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;2191912626 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;2191912626</title>\n",
       "<g id=\"a_node3\"><a xlink:title=\"An exception was raised during the unpacking of &#39;data&#39;: Could not unpack data item with the hinted type &#39;&lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;&#39;. The following packagers were tried to be used to unpack it but raised the exceptions joined:&#10;&#10;Found packager: &#39;PandasDataFramePackager(packable_type=DataFrame, artifact_types=[&#39;dataset&#39;, &#39;file&#39;, &#39;object&#39;, &#39;result&#39;], priority=5)&#39;&#10;Exception: Traceback (most recent call last):&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers_manager.py&quot;, line 660, in _unpack_data_item&#10; &#160;&#160;&#160;return packager.unpack(&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/default_packager.py&quot;, line 237, in unpack&#10; &#160;&#160;&#160;return unpack_method(data_item, **instructions)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/pandas_packagers.py&quot;, line 823, in unpack_file&#10; &#160;&#160;&#160;return formatter.read(file_path=file_path, **read_kwargs)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/pandas_packagers.py&quot;, line 236, in read&#10; &#160;&#160;&#160;obj = pd.read_csv(filepath_or_buffer=file_path, **read_kwargs)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/util/_decorators.py&quot;, line 311, in wrapper&#10; &#160;&#160;&#160;return func(*args, **kwargs)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/parsers/readers.py&quot;, line 678, in read_csv&#10; &#160;&#160;&#160;return _read(filepath_or_buffer, kwds)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/parsers/readers.py&quot;, line 575, in _read&#10; &#160;&#160;&#160;parser = TextFileReader(filepath_or_buffer, **kwds)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/parsers/readers.py&quot;, line 932, in __init__&#10; &#160;&#160;&#160;self._engine = self._make_engine(f, self.engine)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/parsers/readers.py&quot;, line 1216, in _make_engine&#10; &#160;&#160;&#160;self.handles = get_handle( &#160;# type: ignore[call&#45;overload]&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/common.py&quot;, line 786, in get_handle&#10; &#160;&#160;&#160;handle = open(&#10;FileNotFoundError: [Errno 2] No such file or directory: &#39;/User/igz&#45;liveops&#45;demo/data/heart.csv&#39;&#10;&#10;&#10;Found packager: &#39;DefaultPackager(packable_type=Any, artifact_types=[&#39;object&#39;, &#39;result&#39;], priority=Ellipsis)&#39;&#10;Exception: Traceback (most recent call last):&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers_manager.py&quot;, line 684, in _unpack_data_item&#10; &#160;&#160;&#160;return self._default_packager.unpack(&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/default_packager.py&quot;, line 237, in unpack&#10; &#160;&#160;&#160;return unpack_method(data_item, **instructions)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/default_packager.py&quot;, line 353, in unpack_object&#10; &#160;&#160;&#160;return Pickler.unpickle(&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/utils/_pickler.py&quot;, line 158, in unpickle&#10; &#160;&#160;&#160;with open(pickle_path, &quot;rb&quot;) as pickle_file:&#10;FileNotFoundError: [Errno 2] No such file or directory: &#39;/User/igz&#45;liveops&#45;demo/data/heart.csv&#39;&#10;&#10;, caused by: Could not unpack data item with the hinted type &#39;&lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;&#39;. The following packagers were tried to be used to unpack it but raised the exceptions joined:&#10;&#10;Found packager: &#39;PandasDataFramePackager(packable_type=DataFrame, artifact_types=[&#39;dataset&#39;, &#39;file&#39;, &#39;object&#39;, &#39;result&#39;], priority=5)&#39;&#10;Exception: Traceback (most recent call last):&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers_manager.py&quot;, line 660, in _unpack_data_item&#10; &#160;&#160;&#160;return packager.unpack(&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/default_packager.py&quot;, line 237, in unpack&#10; &#160;&#160;&#160;return unpack_method(data_item, **instructions)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/pandas_packagers.py&quot;, line 823, in unpack_file&#10; &#160;&#160;&#160;return formatter.read(file_path=file_path, **read_kwargs)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/pandas_packagers.py&quot;, line 236, in read&#10; &#160;&#160;&#160;obj = pd.read_csv(filepath_or_buffer=file_path, **read_kwargs)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/util/_decorators.py&quot;, line 311, in wrapper&#10; &#160;&#160;&#160;return func(*args, **kwargs)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/parsers/readers.py&quot;, line 678, in read_csv&#10; &#160;&#160;&#160;return _read(filepath_or_buffer, kwds)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/parsers/readers.py&quot;, line 575, in _read&#10; &#160;&#160;&#160;parser = TextFileReader(filepath_or_buffer, **kwds)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/parsers/readers.py&quot;, line 932, in __init__&#10; &#160;&#160;&#160;self._engine = self._make_engine(f, self.engine)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/parsers/readers.py&quot;, line 1216, in _make_engine&#10; &#160;&#160;&#160;self.handles = get_handle( &#160;# type: ignore[call&#45;overload]&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/pandas/io/common.py&quot;, line 786, in get_handle&#10; &#160;&#160;&#160;handle = open(&#10;FileNotFoundError: [Errno 2] No such file or directory: &#39;/User/igz&#45;liveops&#45;demo/data/heart.csv&#39;&#10;&#10;&#10;Found packager: &#39;DefaultPackager(packable_type=Any, artifact_types=[&#39;object&#39;, &#39;result&#39;], priority=Ellipsis)&#39;&#10;Exception: Traceback (most recent call last):&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers_manager.py&quot;, line 684, in _unpack_data_item&#10; &#160;&#160;&#160;return self._default_packager.unpack(&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/default_packager.py&quot;, line 237, in unpack&#10; &#160;&#160;&#160;return unpack_method(data_item, **instructions)&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/packagers/default_packager.py&quot;, line 353, in unpack_object&#10; &#160;&#160;&#160;return Pickler.unpickle(&#10; &#160;File &quot;/usr/local/lib/python3.9/site&#45;packages/mlrun/package/utils/_pickler.py&quot;, line 158, in unpickle&#10; &#160;&#160;&#160;with open(pickle_path, &quot;rb&quot;) as pickle_file:&#10;FileNotFoundError: [Errno 2] No such file or directory: &#39;/User/igz&#45;liveops&#45;demo/data/heart.csv&#39;&#10;&#10;\">\n",
       "<ellipse fill=\"red\" stroke=\"black\" cx=\"273.05\" cy=\"-306\" rx=\"72.59\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"273.05\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">data&#45;get&#45;data</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;826146055 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;826146055</title>\n",
       "<ellipse fill=\"lightgrey\" stroke=\"black\" cx=\"50.05\" cy=\"-234\" rx=\"50.09\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"50.05\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">describe</text>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;2191912626&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;826146055 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;2191912626&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;826146055</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M229.78,-291.42C191.37,-279.36 135.39,-261.79 96.14,-249.47\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"97.14,-246.11 86.55,-246.46 95.04,-252.79 97.14,-246.11\"/>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;3207923530 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;3207923530</title>\n",
       "<ellipse fill=\"lightgrey\" stroke=\"black\" cx=\"273.05\" cy=\"-234\" rx=\"154.87\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"273.05\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">validate&#45;validate&#45;data&#45;integrity</text>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;2191912626&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;3207923530 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;2191912626&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;3207923530</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M273.05,-287.7C273.05,-279.98 273.05,-270.71 273.05,-262.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"276.55,-262.1 273.05,-252.1 269.55,-262.1 276.55,-262.1\"/>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017</title>\n",
       "<ellipse fill=\"lightgrey\" stroke=\"black\" cx=\"364.05\" cy=\"-162\" rx=\"92.88\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"364.05\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">data&#45;process&#45;data</text>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;2191912626&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;2191912626&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M343.15,-300.9C377.31,-294.94 415.57,-281.52 437.05,-252 446.46,-239.06 444.31,-230.25 437.05,-216 430.46,-203.09 419.1,-192.59 407.27,-184.44\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"408.95,-181.36 398.64,-178.92 405.18,-187.25 408.95,-181.36\"/>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;3207923530&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;3207923530&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M295.07,-216.05C306.9,-206.96 321.63,-195.63 334.44,-185.77\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"336.74,-188.42 342.53,-179.55 332.47,-182.87 336.74,-188.42\"/>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;14175956 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;14175956</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M329.45,-145.08C314.79,-136.37 299.23,-124.05 291.05,-108 283.78,-93.75 283.78,-86.25 291.05,-72 299.29,-55.82 315.04,-43.43 329.81,-34.71\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"331.73,-37.64 338.8,-29.75 328.35,-31.51 331.73,-37.64\"/>\n",
       "</g>\n",
       "<!-- gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;1947655188 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>gitops&#45;training&#45;pipeline&#45;c6srb&#45;549906017&#45;&gt;gitops&#45;training&#45;pipeline&#45;c6srb&#45;1947655188</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M386.09,-144.41C398.07,-135.39 413.09,-124.09 426.21,-114.21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"428.63,-116.77 434.52,-107.96 424.42,-111.17 428.63,-116.77\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7f6778bbd790>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<h2>Run Results</h2><h3>[info] Workflow 14715a49-ad72-40c5-a584-140fc265e747 finished with 1 errors</h3><br>click the hyper links below to see detailed results<br><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>uid</th>\n",
       "      <th>start</th>\n",
       "      <th>state</th>\n",
       "      <th>name</th>\n",
       "      <th>parameters</th>\n",
       "      <th>results</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><div title=\"04291b3e26ad4b7c88e01d4d04f3d9df\"><a href=\"https://dashboard.default-tenant.app.us-tmp-353.iguazio-cd2.com/mlprojects/cicd-flow/jobs/monitor/04291b3e26ad4b7c88e01d4d04f3d9df/overview\" target=\"_blank\" >...04f3d9df</a></div></td>\n",
       "      <td>Sep 18 23:37:56</td>\n",
       "      <td><div style=\"color: red;\" title=\"An exception was raised during the unpacking of 'data': Could not unpack data item with the hinted type '<class 'pandas.core.frame.DataFrame'>'. The following packagers were tried to be used to unpack it but raised the exceptions joined:\\n\\nFound packager: 'PandasDataFramePackager(packable_type=DataFrame, artifact_types=['dataset', 'file', 'object', 'result'], priority=5)'\\nException: Traceback (most recent call last):\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers_manager.py\", line 660, in _unpack_data_item\\n    return packager.unpack(\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/default_packager.py\", line 237, in unpack\\n    return unpack_method(data_item, **instructions)\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/pandas_packagers.py\", line 823, in unpack_file\\n    return formatter.read(file_path=file_path, **read_kwargs)\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/pandas_packagers.py\", line 236, in read\\n    obj = pd.read_csv(filepath_or_buffer=file_path, **read_kwargs)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/util/_decorators.py\", line 311, in wrapper\\n    return func(*args, **kwargs)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 678, in read_csv\\n    return _read(filepath_or_buffer, kwds)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 575, in _read\\n    parser = TextFileReader(filepath_or_buffer, **kwds)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 932, in __init__\\n    self._engine = self._make_engine(f, self.engine)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 1216, in _make_engine\\n    self.handles = get_handle(  # type: ignore[call-overload]\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/common.py\", line 786, in get_handle\\n    handle = open(\\nFileNotFoundError: [Errno 2] No such file or directory: '/User/igz-liveops-demo/data/heart.csv'\\n\\n\\nFound packager: 'DefaultPackager(packable_type=Any, artifact_types=['object', 'result'], priority=Ellipsis)'\\nException: Traceback (most recent call last):\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers_manager.py\", line 684, in _unpack_data_item\\n    return self._default_packager.unpack(\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/default_packager.py\", line 237, in unpack\\n    return unpack_method(data_item, **instructions)\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/default_packager.py\", line 353, in unpack_object\\n    return Pickler.unpickle(\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/utils/_pickler.py\", line 158, in unpickle\\n    with open(pickle_path, \"rb\") as pickle_file:\\nFileNotFoundError: [Errno 2] No such file or directory: '/User/igz-liveops-demo/data/heart.csv'\\n\\n, caused by: Could not unpack data item with the hinted type '<class 'pandas.core.frame.DataFrame'>'. The following packagers were tried to be used to unpack it but raised the exceptions joined:\\n\\nFound packager: 'PandasDataFramePackager(packable_type=DataFrame, artifact_types=['dataset', 'file', 'object', 'result'], priority=5)'\\nException: Traceback (most recent call last):\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers_manager.py\", line 660, in _unpack_data_item\\n    return packager.unpack(\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/default_packager.py\", line 237, in unpack\\n    return unpack_method(data_item, **instructions)\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/pandas_packagers.py\", line 823, in unpack_file\\n    return formatter.read(file_path=file_path, **read_kwargs)\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/pandas_packagers.py\", line 236, in read\\n    obj = pd.read_csv(filepath_or_buffer=file_path, **read_kwargs)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/util/_decorators.py\", line 311, in wrapper\\n    return func(*args, **kwargs)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 678, in read_csv\\n    return _read(filepath_or_buffer, kwds)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 575, in _read\\n    parser = TextFileReader(filepath_or_buffer, **kwds)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 932, in __init__\\n    self._engine = self._make_engine(f, self.engine)\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 1216, in _make_engine\\n    self.handles = get_handle(  # type: ignore[call-overload]\\n  File \"/usr/local/lib/python3.9/site-packages/pandas/io/common.py\", line 786, in get_handle\\n    handle = open(\\nFileNotFoundError: [Errno 2] No such file or directory: '/User/igz-liveops-demo/data/heart.csv'\\n\\n\\nFound packager: 'DefaultPackager(packable_type=Any, artifact_types=['object', 'result'], priority=Ellipsis)'\\nException: Traceback (most recent call last):\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers_manager.py\", line 684, in _unpack_data_item\\n    return self._default_packager.unpack(\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/default_packager.py\", line 237, in unpack\\n    return unpack_method(data_item, **instructions)\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/packagers/default_packager.py\", line 353, in unpack_object\\n    return Pickler.unpickle(\\n  File \"/usr/local/lib/python3.9/site-packages/mlrun/package/utils/_pickler.py\", line 158, in unpickle\\n    with open(pickle_path, \"rb\") as pickle_file:\\nFileNotFoundError: [Errno 2] No such file or directory: '/User/igz-liveops-demo/data/heart.csv'\\n\\n\">error</div></td>\n",
       "      <td>data-get-data</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "RuntimeError",
     "evalue": "Pipeline run status Failed",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m run_id \u001b[38;5;241m=\u001b[39m \u001b[43mproject\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[43m    \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtrain\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m    \u001b[49m\u001b[43marguments\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrain_config\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdict\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdirty\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m    \u001b[49m\u001b[43mwatch\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\n\u001b[1;32m      6\u001b[0m \u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pythonlibs/mlrun-extended/lib/python3.9/site-packages/mlrun/projects/project.py:2347\u001b[0m, in \u001b[0;36mMlrunProject.run\u001b[0;34m(self, name, workflow_path, arguments, artifact_path, workflow_handler, namespace, sync, watch, dirty, ttl, engine, local, schedule, timeout, overwrite, source, cleanup_ttl)\u001b[0m\n\u001b[1;32m   2345\u001b[0m workflow_spec\u001b[38;5;241m.\u001b[39mclear_tmp()\n\u001b[1;32m   2346\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (timeout \u001b[38;5;129;01mor\u001b[39;00m watch) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m workflow_spec\u001b[38;5;241m.\u001b[39mschedule:\n\u001b[0;32m-> 2347\u001b[0m     \u001b[43mrun\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_run_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43mproject\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2348\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m run\n",
      "File \u001b[0;32m~/.pythonlibs/mlrun-extended/lib/python3.9/site-packages/mlrun/projects/pipelines.py:668\u001b[0m, in \u001b[0;36m_KFPRunner.get_run_status\u001b[0;34m(project, run, timeout, expected_statuses, notifiers)\u001b[0m\n\u001b[1;32m    665\u001b[0m notifiers\u001b[38;5;241m.\u001b[39mpush(text, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minfo\u001b[39m\u001b[38;5;124m\"\u001b[39m, runs)\n\u001b[1;32m    667\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m raise_error:\n\u001b[0;32m--> 668\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m raise_error\n\u001b[1;32m    669\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m state, had_errors, text\n",
      "File \u001b[0;32m~/.pythonlibs/mlrun-extended/lib/python3.9/site-packages/mlrun/projects/pipelines.py:643\u001b[0m, in \u001b[0;36m_KFPRunner.get_run_status\u001b[0;34m(project, run, timeout, expected_statuses, notifiers)\u001b[0m\n\u001b[1;32m    641\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout:\n\u001b[1;32m    642\u001b[0m         logger\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwaiting for pipeline run completion\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 643\u001b[0m         state \u001b[38;5;241m=\u001b[39m \u001b[43mrun\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait_for_completion\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    644\u001b[0m \u001b[43m            \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexpected_statuses\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexpected_statuses\u001b[49m\n\u001b[1;32m    645\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    646\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    647\u001b[0m     \u001b[38;5;66;03m# push runs table also when we have errors\u001b[39;00m\n\u001b[1;32m    648\u001b[0m     raise_error \u001b[38;5;241m=\u001b[39m exc\n",
      "File \u001b[0;32m~/.pythonlibs/mlrun-extended/lib/python3.9/site-packages/mlrun/projects/pipelines.py:461\u001b[0m, in \u001b[0;36m_PipelineRunStatus.wait_for_completion\u001b[0;34m(self, timeout, expected_statuses)\u001b[0m\n\u001b[1;32m    460\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwait_for_completion\u001b[39m(\u001b[38;5;28mself\u001b[39m, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, expected_statuses\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 461\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_state \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait_for_completion\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    462\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    463\u001b[0m \u001b[43m        \u001b[49m\u001b[43mproject\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mproject\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    464\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    465\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexpected_statuses\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexpected_statuses\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    466\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    467\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_state\n",
      "File \u001b[0;32m~/.pythonlibs/mlrun-extended/lib/python3.9/site-packages/mlrun/projects/pipelines.py:609\u001b[0m, in \u001b[0;36m_KFPRunner.wait_for_completion\u001b[0;34m(run_id, project, timeout, expected_statuses)\u001b[0m\n\u001b[1;32m    607\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m60\u001b[39m \u001b[38;5;241m*\u001b[39m \u001b[38;5;241m60\u001b[39m\n\u001b[1;32m    608\u001b[0m project_name \u001b[38;5;241m=\u001b[39m project\u001b[38;5;241m.\u001b[39mmetadata\u001b[38;5;241m.\u001b[39mname \u001b[38;5;28;01mif\u001b[39;00m project \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m--> 609\u001b[0m run_info \u001b[38;5;241m=\u001b[39m \u001b[43mwait_for_pipeline_completion\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    610\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrun_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    611\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    612\u001b[0m \u001b[43m    \u001b[49m\u001b[43mexpected_statuses\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexpected_statuses\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    613\u001b[0m \u001b[43m    \u001b[49m\u001b[43mproject\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mproject_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    614\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    615\u001b[0m status \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    616\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m run_info:\n",
      "File \u001b[0;32m~/.pythonlibs/mlrun-extended/lib/python3.9/site-packages/mlrun/run.py:1193\u001b[0m, in \u001b[0;36mwait_for_pipeline_completion\u001b[0;34m(run_id, timeout, expected_statuses, namespace, remote, project)\u001b[0m\n\u001b[1;32m   1191\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m expected_statuses:\n\u001b[1;32m   1192\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m expected_statuses:\n\u001b[0;32m-> 1193\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[1;32m   1194\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPipeline run status \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstatus\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;250m \u001b[39m\u001b[38;5;241m+\u001b[39m\u001b[38;5;250m \u001b[39mmessage\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mif\u001b[39;00m\u001b[38;5;250m \u001b[39mmessage\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01melse\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1195\u001b[0m         )\n\u001b[1;32m   1197\u001b[0m logger\u001b[38;5;241m.\u001b[39mdebug(\n\u001b[1;32m   1198\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFinished waiting for pipeline completion.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1199\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m run_id: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mrun_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1202\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m namespace: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnamespace\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1203\u001b[0m )\n\u001b[1;32m   1205\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Pipeline run status Failed"
     ]
    }
   ],
   "source": [
    "run_id = project.run(\n",
    "    name=\"train\",\n",
    "    arguments=train_config.dict(),\n",
    "    dirty=True,\n",
    "    watch=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Evaluation Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# model_path = project.list_models()\n",
    "# train_set = ...\n",
    "# test_set = ...\n",
    "\n",
    "evaluate_config = EvaluateConfig("
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Deployment Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Invoke Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> 2023-08-17 21:06:26,931 [info] invoking function: {'method': 'POST', 'path': 'http://nuclio-liveops-demo-serving.default-tenant.svc.cluster.local:8080/v2/models/model/infer'}\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "bad function response 400: ValueError: X has 4 features, but RandomForestClassifier is expecting 23 features as input.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m serving_fn \u001b[38;5;241m=\u001b[39m project\u001b[38;5;241m.\u001b[39mget_function(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mserving\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 2\u001b[0m \u001b[43mserving_fn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minvoke\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m/v2/models/model/infer\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43minputs\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m6.1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m2.8\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m4.7\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1.2\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m5.7\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m3.8\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1.7\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0.3\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m7.7\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m2.6\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m6.9\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m2.3\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pythonlibs/mlrun-base/lib/python3.9/site-packages/mlrun/runtimes/function.py:913\u001b[0m, in \u001b[0;36mRemoteRuntime.invoke\u001b[0;34m(self, path, body, method, headers, dashboard, force_external_address, auth_info, mock)\u001b[0m\n\u001b[1;32m    909\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m(\n\u001b[1;32m    910\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124merror: cannot run function at url \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpath\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m, \u001b[39m\u001b[38;5;132;01m{\u001b[39;00merr_to_str(err)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    911\u001b[0m     )\n\u001b[1;32m    912\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m resp\u001b[38;5;241m.\u001b[39mok:\n\u001b[0;32m--> 913\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbad function response \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresp\u001b[38;5;241m.\u001b[39mstatus_code\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresp\u001b[38;5;241m.\u001b[39mtext\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    915\u001b[0m data \u001b[38;5;241m=\u001b[39m resp\u001b[38;5;241m.\u001b[39mcontent\n\u001b[1;32m    916\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m resp\u001b[38;5;241m.\u001b[39mheaders[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent-type\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapplication/json\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "\u001b[0;31mRuntimeError\u001b[0m: bad function response 400: ValueError: X has 4 features, but RandomForestClassifier is expecting 23 features as input."
     ]
    }
   ],
   "source": [
    "serving_fn = project.get_function(\"serving\")\n",
    "serving_fn.invoke('/v2/models/model/infer', {\"inputs\" : [[6.1, 2.8, 4.7, 1.2], [5.7, 3.8, 1.7, 0.3], [7.7, 2.6, 6.9, 2.3]]})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Simulate Traffic\n",
    "\n",
    "MLRun real-time drift detection requires 10,000 events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get serving function\n",
    "serving_fn = project.get_function(\"serving\")\n",
    "model_name = \"model\"\n",
    "\n",
    "# Get test set\n",
    "df = project.get_artifact(\"train_test_set\").to_dataitem().as_df()\n",
    "display(df.head())\n",
    "\n",
    "# Transform to list\n",
    "# Get training set as list\n",
    "iris_data = df.drop(\"label\", axis=1).to_dict(orient=\"split\")[\"data\"]\n",
    "iris_data[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import logging\n",
    "from random import choice\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "# Suppress print messages\n",
    "logging.getLogger(name=\"mlrun\").setLevel(logging.WARNING)\n",
    "\n",
    "# Simulate traffic using random elements from training set\n",
    "for i in tqdm(range(12_000)):\n",
    "    data_point = choice(iris_data)\n",
    "    serving_fn.invoke(\n",
    "        f\"v2/models/{model_name}/infer\", json.dumps({\"inputs\": [data_point]})\n",
    "    )\n",
    "\n",
    "# Resume normal logging\n",
    "logging.getLogger(name=\"mlrun\").setLevel(logging.INFO)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mlrun-extended",
   "language": "python",
   "name": "conda-env-mlrun-extended-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
